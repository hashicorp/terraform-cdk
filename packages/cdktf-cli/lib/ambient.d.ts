declare module 'ink-confirm-input';