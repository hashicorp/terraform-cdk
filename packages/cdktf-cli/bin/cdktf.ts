import * as yargs from 'yargs';
import * as semver from 'semver';

if (semver.lt(process.version, '10.12.0')) { console.error("Need at least Node v10.12 to run") ; process.exit(1) }

const args = yargs
  .commandDir('cmds')
  .recommendCommands()
  .wrap(yargs.terminalWidth())
  .showHelpOnFail(false)
  .env('CDKTF')
  .epilogue(`Options can be specified via environment variables with the "CDKTF_" prefix (e.g. "CDKTF_OUTPUT")`)
  .help()
  .alias('h', 'help')
  .argv;

if (args._.length === 0) {
  yargs.showHelp();
}
