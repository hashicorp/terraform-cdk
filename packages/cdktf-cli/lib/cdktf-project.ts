// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { AbortController } from "node-abort-controller"; // polyfill until we update to node 16
import { SynthesizedStack, SynthStack } from "./synth-stack";
import { printAnnotations } from "./synth";
import { CdktfStack, StackApprovalUpdate, StackUpdate } from "./cdktf-stack";
import { Errors } from "./errors";
import { TerraformPlan } from "./models/terraform";
import { NestedTerraformOutputs } from "./output";
import { logger } from "./logging";
import minimatch from "minimatch";
import { createEnhanceLogMessage } from "./execution-logs";
import { ensureAllSettledBeforeThrowing } from "./util";

type MultiStackApprovalUpdate = {
  type: "waiting for approval";
  stackName: string;
  plan: TerraformPlan;
  approve: () => void;
  dismiss: () => void;
  stop: () => void;
};

export type ProjectUpdate =
  | {
      type: "synthesizing";
    }
  | {
      type: "synthesized";
      stacks: SynthesizedStack[];
      errorMessage?: string;
    }
  | StackUpdate
  | MultiStackApprovalUpdate;

function getSingleStack(
  stacks: SynthesizedStack[],
  stackName?: string,
  targetAction?: string
) {
  if (!stacks) {
    throw Errors.Internal(
      "Trying to access a stack before it has been synthesized"
    );
  }

  if (stackName) {
    const stack = stacks.find((s) => s.name === stackName);
    if (!stack) {
      throw Errors.Usage("Could not find stack: " + stackName);
    }

    return stack;
  }

  if (stacks.length === 1) {
    return stacks[0];
  }

  throw Errors.Usage(
    `Found more than one stack, please specify a target stack. Run cdktf ${
      targetAction || "<verb>"
    } <stack> with one of these stacks: ${stacks
      .map((s) => s.name)
      .join(", ")} `
  );
}

export function getMultipleStacks(
  stacks: SynthesizedStack[],
  patterns?: string[],
  targetAction?: string
) {
  if (!patterns || !patterns.length) {
    if (stacks.length === 1) {
      return [stacks[0]];
    }
    throw Errors.Usage(
      `Found more than one stack, please specify a target stack. Run cdktf ${
        targetAction || "<verb>"
      } <stack> with one of these stacks: ${stacks
        .map((s) => s.name)
        .join(", ")} `
    );
  }

  return patterns.flatMap((pattern) => {
    const matchingStacks = stacks.filter((stack) =>
      minimatch(stack.name, pattern)
    );

    if (matchingStacks.length === 0) {
      throw Errors.Usage(`Could not find stack for pattern '${pattern}'`);
    }

    return matchingStacks;
  });
}

// Returns the first stack that has no unmet dependencies
// An unmet dependency is a dependency that has not been deployed yet
// If there is no unfinished stack, returns undefined
// If there is no stack ready to be worked on, it returns a promise that will resolve as soon as there is a follow-up stack available
export async function getStackWithNoUnmetDependencies(
  stackExecutors: CdktfStack[]
): Promise<CdktfStack | undefined> {
  logger.debug("Checking for stacks with unmet dependencies");
  logger.debug("stack executors:", stackExecutors);
  const pendingStacks = stackExecutors.filter((executor) => executor.isPending);
  logger.debug("pending stacks:", stackExecutors);

  if (pendingStacks.length === 0) {
    return undefined;
  }

  const currentlyReadyStack = pendingStacks.find((executor) =>
    executor.stack.dependencies.every(
      (dependency) =>
        stackExecutors.find((executor) => executor.stack.name === dependency)
          ?.currentState === "done"
    )
  );

  if (currentlyReadyStack) {
    logger.debug("Found a stack with no unmet dependencies");
    return currentlyReadyStack;
  }

  const stackExecutionPromises = stackExecutors
    .filter((ex) => ex.currentWorkPromise)
    .map((ex) => ex.currentWorkPromise);

  logger.debug(
    `${stackExecutionPromises.length} stacks are currently busy, waiting for one to finish`
  );

  if (!stackExecutionPromises.length) {
    return undefined;
  }

  await ensureAllSettledBeforeThrowing(
    Promise.race(stackExecutionPromises),
    stackExecutionPromises
  );

  return await getStackWithNoUnmetDependencies(stackExecutors);
}

function findAllDependantStacks(
  stackExecutors: CdktfStack[],
  stackName: string
): CdktfStack[] {
  return stackExecutors.filter((innerExecutor) =>
    innerExecutor.stack.dependencies.includes(stackName)
  );
}

function findAllNestedDependantStacks(
  stackExecutors: CdktfStack[],
  stackName: string,
  knownDependantStackNames: Set<string> = new Set()
): CdktfStack[] {
  const dependantStacks = findAllDependantStacks(stackExecutors, stackName);
  dependantStacks.forEach((stack) => {
    if (knownDependantStackNames.has(stack.stack.name)) {
      return;
    }

    knownDependantStackNames.add(stack.stack.name);
    findAllNestedDependantStacks(
      stackExecutors,
      stack.stack.name,
      knownDependantStackNames
    );
  });

  return stackExecutors.filter((executor) =>
    knownDependantStackNames.has(executor.stack.name)
  );
}

// Returns the first stack that has no dependents that need to be destroyed first
async function getStackWithNoUnmetDependants(
  stackExecutors: CdktfStack[]
): Promise<CdktfStack | undefined> {
  logger.debug("Checking for stacks with unmet dependants");
  logger.debug("stack executors:", stackExecutors);
  const pendingStacks = stackExecutors.filter((executor) => executor.isPending);
  logger.debug("pending stacks:", stackExecutors);

  if (pendingStacks.length === 0) {
    return undefined;
  }

  const currentlyReadyStack = pendingStacks.find((executor) => {
    const dependantStacks = findAllDependantStacks(
      stackExecutors,
      executor.stack.name
    );
    return dependantStacks.every((stack) => stack.currentState === "done");
  });

  if (currentlyReadyStack) {
    logger.debug("Found a stack with no unmet dependants");
    return currentlyReadyStack;
  }
  const stackExecutionPromises = stackExecutors
    .filter((ex) => ex.currentWorkPromise)
    .map((ex) => ex.currentWorkPromise);

  logger.debug(
    `${stackExecutionPromises.length} stacks are currently busy, waiting for one to finish`
  );
  if (!stackExecutionPromises.length) {
    return undefined;
  }

  await Promise.race(stackExecutionPromises);
  return await getStackWithNoUnmetDependants(stackExecutors);
}

// Throws an error if there is a dependant stack that is not included
function checkIfAllDependantsAreIncluded(
  stacksToRun: SynthesizedStack[],
  allStacks: SynthesizedStack[]
) {
  const allDependants = new Set<string>();
  stacksToRun
    .map((stack) =>
      allStacks.filter((s) => s.dependencies.includes(stack.name))
    )
    .flat()
    .forEach((dependant) => allDependants.add(dependant.name));

  const stackNames = stacksToRun.map((stack) => stack.name);
  const missingDependants = [...allDependants].filter(
    (dependant) => !stackNames.includes(dependant)
  );

  if (missingDependants.length > 0) {
    throw Errors.Usage(
      `The following dependant stacks are not included in the stacks to run: ${missingDependants.join(
        ", "
      )}. Either add them or add the --ignore-missing-stack-dependencies flag.`
    );
  }
}

// Throws an error if there is a dependency that is not included
// Cycles are detected on dependency creation at synthesis time
// Running this prevents us from being in a situation where we have to wait for a stack to be deployed
// that is not included to be run
function checkIfAllDependenciesAreIncluded(stacksToRun: SynthesizedStack[]) {
  const allDependencies = new Set<string>();
  stacksToRun
    .map((stack) => stack.dependencies)
    .flat()
    .forEach((dependency) => allDependencies.add(dependency));

  const stackNames = stacksToRun.map((stack) => stack.name);
  const missingDependencies = [...allDependencies].filter(
    (dependency) => !stackNames.includes(dependency)
  );

  if (missingDependencies.length > 0) {
    throw Errors.Usage(
      `The following dependencies are not included in the stacks to run: ${missingDependencies.join(
        ", "
      )}. Either add them or add the --ignore-missing-stack-dependencies flag.`
    );
  }
}

export type SingleStackOptions = {
  stackName?: string;
};

export type MultipleStackOptions = {
  stackNames?: string[];
};

export type ExecutionOptions = MultipleStackOptions & {
  autoApprove?: boolean;
  ignoreMissingStackDependencies?: boolean;
  parallelism?: number;
  refreshOnly?: boolean;
};

export type DiffOptions = SingleStackOptions & {
  refreshOnly?: boolean;
};

type LogMessage = {
  stackName: string;
  messageWithConstructPath?: string;
  message: string;
};

// Stores a log value of a certain type until it can be sent
type Buffered<T, V> = {
  cb: (item: T) => void;
  value: T;
  type: V;
};

export type CdktfProjectOptions = {
  synthCommand: string;
  outDir: string;
  onUpdate: (update: ProjectUpdate) => void;
  onLog?: (log: LogMessage) => void;
  workingDirectory?: string;
};
export class CdktfProject {
  public stacks?: SynthesizedStack[];
  public hardAbort: () => void;

  private synthCommand: string;
  private outDir: string;
  private workingDirectory: string;
  private onUpdate: (update: ProjectUpdate) => void;
  private onLog?: (log: LogMessage) => void;
  private abortSignal: AbortSignal;

  // Set during deploy / destroy
  public stacksToRun: CdktfStack[] = [];
  // This means sth different in deploy / destroy
  private stopAllStacksThatCanNotRunWithout: (stackName: string) => void =
    () => {}; // eslint-disable-line @typescript-eslint/no-empty-function
  // Pauses all progress / status events from being forwarded to the user
  // If set from true to false, the events will be sent through the channels they came in
  // (until a waiting for approval event is sent)
  private waitingForApproval = false;
  private eventBuffer: Array<
    | Buffered<ProjectUpdate, "projectUpdate">
    | Buffered<LogMessage, "logMessage">
  > = [];

  constructor({
    synthCommand,
    outDir,
    onUpdate,
    onLog,
    workingDirectory = process.cwd(),
  }: CdktfProjectOptions) {
    this.synthCommand = synthCommand;
    this.outDir = outDir;
    this.workingDirectory = workingDirectory;
    this.onUpdate = onUpdate;
    this.onLog = onLog;
    const ac = new AbortController();
    this.abortSignal = ac.signal;

    this.hardAbort = ac.abort.bind(ac);
  }

  private stopAllStacks() {
    this.stacksToRun.forEach((stack) => stack.stop());
    this.eventBuffer = this.eventBuffer.filter(
      (event) =>
        event.type === "projectUpdate"
          ? event.value.type !== "waiting for approval" // we want to filter out the waiting for approval events
          : true // we want all other types
    );
  }

  private waitForApproval() {
    this.waitingForApproval = true;
  }

  private resumeAfterApproval() {
    // We first need to flush all events, we can not resume if there is a new waiting for approval update
    let event = this.eventBuffer.shift();
    while (event) {
      if (event.type === "projectUpdate") {
        event.cb(event.value);
        if (event.value.type === "waiting for approval") {
          // We have to wait for approval again, therefore we can not resume
          return;
        }
      }
      if (event.type === "logMessage") {
        event.cb(event.value);
      }

      event = this.eventBuffer.shift();
    }

    // If we reach this point there was no waiting for approval event, so we can safely resume
    this.waitingForApproval = false;
  }

  private handleApprovalProcess(cb: (updateToSend: ProjectUpdate) => void) {
    return (update: StackUpdate | StackApprovalUpdate) => {
      const bufferCb = (bufferedUpdate: ProjectUpdate) => {
        this.eventBuffer.push({
          cb,
          value: bufferedUpdate,
          type: "projectUpdate",
        });
      };
      const bufferableCb = this.waitingForApproval ? bufferCb : cb;

      if (update.type === "waiting for stack approval") {
        const callbacks = (update: StackApprovalUpdate) => ({
          approve: () => {
            update.approve();
            this.resumeAfterApproval();
          },
          dismiss: () => {
            update.reject();

            this.stopAllStacksThatCanNotRunWithout(update.stackName);
            this.resumeAfterApproval();
          },
          stop: () => {
            update.reject();
            this.stopAllStacks();
            this.resumeAfterApproval();
          },
        });
        this.waitForApproval();

        bufferableCb({
          type: "waiting for approval",
          stackName: update.stackName,
          plan: update.plan,
          ...callbacks(update),
        });
      } else {
        bufferableCb(update);
      }
    };
  }

  private bufferWhileWaitingForApproval(cb?: (msg: LogMessage) => void) {
    if (!cb) {
      return undefined;
    }

    return (msg: LogMessage) => {
      if (this.waitingForApproval) {
        this.eventBuffer.push({
          cb,
          value: msg,
          type: "logMessage",
        });
      } else {
        cb(msg);
      }
    };
  }

  public getStackExecutor(
    stack: SynthesizedStack,
    opts: ExecutionOptions = {}
  ) {
    const enhanceLogMessage = createEnhanceLogMessage(stack);
    const onLog = this.bufferWhileWaitingForApproval(this.onLog);
    return new CdktfStack({
      ...opts,
      stack,
      onUpdate: this.handleApprovalProcess(this.onUpdate),
      onLog: onLog
        ? ({ message }) =>
            onLog({
              stackName: stack.name,
              message,
              messageWithConstructPath: enhanceLogMessage(message),
            })
        : undefined,
      abortSignal: this.abortSignal,
    });
  }

  public get outputsByConstructId() {
    return this.stacksToRun.reduce(
      (acc, stack) => ({
        ...acc,
        ...stack.outputsByConstructId,
      }),
      {} as NestedTerraformOutputs
    );
  }

  public async synth() {
    this.onUpdate({
      type: "synthesizing",
    });
    const stacks = await SynthStack.synth(
      this.abortSignal,
      this.synthCommand,
      this.outDir,
      this.workingDirectory
    );

    printAnnotations(stacks);

    this.onUpdate({
      type: "synthesized",
      stacks,
    });

    return stacks;
  }

  public async diff(opts?: DiffOptions) {
    const stacks = await this.synth();
    const stack = this.getStackExecutor(
      getSingleStack(stacks, opts?.stackName, "diff")
    );
    await stack.diff({ refreshOnly: opts?.refreshOnly });
    if (!stack.currentPlan)
      throw Errors.External(
        `Stack failed to plan: ${stack.stack.name}. Please check the logs for more information.`
      );
    return stack.currentPlan;
  }

  private async execute(
    method: "deploy" | "destroy",
    next: () => Promise<CdktfStack | undefined>,
    parallelism: number,
    refreshOnly?: boolean
  ) {
    // We only support refresh only on deploy, a bit of a leaky abstraction here
    if (refreshOnly && method !== "deploy") {
      throw Errors.Internal(`Refresh only is only supported on deploy`);
    }

    const maxParallelRuns = parallelism === -1 ? Infinity : parallelism;
    while (this.stacksToRun.filter((stack) => stack.isPending).length > 0) {
      const runningStacks = this.stacksToRun.filter((stack) => stack.isRunning);
      if (runningStacks.length >= maxParallelRuns) {
        await Promise.race(runningStacks.map((s) => s.currentWorkPromise));
      } else {
        try {
          const nextRunningExecutor = await next();
          if (!nextRunningExecutor) {
            // In this case we have no pending stacks, but we also can not find a new executor
            break;
          }
          method === "deploy"
            ? nextRunningExecutor.deploy(refreshOnly)
            : nextRunningExecutor.destroy();
        } catch (e) {
          // await next() threw an error because a stack failed to apply/destroy
          // wait for all other currently running stacks to complete before propagating that error
          logger.debug(
            "Encountered an error while awaiting stack to finish",
            e
          );
          const openStacks = this.stacksToRun.filter(
            (ex) => ex.currentWorkPromise
          );
          logger.debug(
            "Waiting for still running stacks to finish:",
            openStacks
          );
          await Promise.allSettled(
            openStacks.map((ex) => ex.currentWorkPromise)
          );
          logger.debug(
            "Done waiting for still running stacks. All pending work finished"
          );
          throw e;
        }
      }
    }

    // We end the loop when all stacks are started, now we need to wait for them to be done
    // We wait for all work to finish even if one of the promises threw an error.
    const currentWork = this.stacksToRun
      .filter((ex) => ex.currentWorkPromise)
      .map((ex) => ex.currentWorkPromise);

    await ensureAllSettledBeforeThrowing(Promise.all(currentWork), currentWork);
  }

  public async deploy(opts: ExecutionOptions = {}) {
    const parallelism = opts.parallelism || -1;
    const stacks = await this.synth();
    const stacksToRun = getMultipleStacks(stacks, opts.stackNames, "deploy");
    if (!opts.ignoreMissingStackDependencies) {
      checkIfAllDependenciesAreIncluded(stacksToRun);
    }

    this.stopAllStacksThatCanNotRunWithout = (stackName: string) => {
      findAllNestedDependantStacks(this.stacksToRun, stackName).forEach(
        (stack) => stack.stop()
      );
    };

    this.stacksToRun = stacksToRun.map((stack) =>
      this.getStackExecutor(stack, opts)
    );

    const next = opts.ignoreMissingStackDependencies
      ? () =>
          Promise.resolve(
            this.stacksToRun.filter((stack) => stack.isPending)[0]
          )
      : () => getStackWithNoUnmetDependencies(this.stacksToRun);

    await this.execute("deploy", next, parallelism, opts.refreshOnly);

    const unprocessedStacks = this.stacksToRun.filter(
      (executor) => executor.isPending
    );
    if (unprocessedStacks.length > 0) {
      throw Errors.External(
        `Some stacks failed to deploy: ${unprocessedStacks
          .map((s) => s.stack.name)
          .join(", ")}. Please check the logs for more information.`
      );
    }
  }

  public async destroy(opts: ExecutionOptions = {}) {
    const parallelism = opts.parallelism || -1;
    const stacks = await this.synth();
    const stacksToRun = getMultipleStacks(stacks, opts.stackNames, "destroy");

    if (!opts.ignoreMissingStackDependencies) {
      checkIfAllDependantsAreIncluded(stacksToRun, stacks);
    }

    this.stopAllStacksThatCanNotRunWithout = (stackName: string) => {
      const stackExecutor = this.stacksToRun.find(
        (s) => s.stack.name === stackName
      );
      if (!stackExecutor) {
        throw Errors.Internal(
          `Could not find stack "${stackName}" that was stopped`
        );
      }

      stackExecutor.stack.dependencies.forEach((dependant) => {
        this.stopAllStacksThatCanNotRunWithout(dependant);

        const dependantStack = this.stacksToRun.find(
          (s) => s.stack.name === dependant
        );
        if (!dependantStack) {
          throw Errors.Internal(
            `Could not find stack "${dependant}" that was stopped`
          );
        }

        dependantStack.stop();
      });
    };
    this.stacksToRun = stacksToRun.map((stack) =>
      this.getStackExecutor(stack, opts)
    );
    const next = opts.ignoreMissingStackDependencies
      ? () =>
          Promise.resolve(
            this.stacksToRun.filter((stack) => stack.currentState !== "done")[0]
          )
      : () => getStackWithNoUnmetDependants(this.stacksToRun);

    await this.execute("destroy", next, parallelism);

    const unprocessedStacks = this.stacksToRun.filter(
      (executor) => executor.isPending
    );
    if (unprocessedStacks.length > 0) {
      throw Errors.External(
        `Some stacks failed to destroy: ${unprocessedStacks
          .map((s) => s.stack.name)
          .join(", ")}. Please check the logs for more information.`
      );
    }
  }

  public async fetchOutputs(opts: MultipleStackOptions) {
    const stacks = await this.synth();

    const stacksToRun = getMultipleStacks(
      stacks,
      opts.stackNames || [],
      "deploy"
    );

    if (stacksToRun.length === 0) {
      throw new Error("No stacks to fetch outputs for specified");
    }

    this.stacksToRun = stacksToRun.map((stack) =>
      this.getStackExecutor(stack, opts)
    );

    const outputs = await Promise.all(
      this.stacksToRun.map(async (s) => {
        const output = await s.fetchOutputs();
        return {
          [s.stack.name]: output,
        };
      })
    );

    return outputs.reduce(
      (acc, curr) => ({ ...acc, ...curr }),
      {}
    ) as NestedTerraformOutputs;
  }
}
