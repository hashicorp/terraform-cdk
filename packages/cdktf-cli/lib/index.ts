// Copyright (c) HashiCorp, Inc
// Licensed under the Mozilla Public License 2.0
// This is the programmatic entrypoint that the CLI uses.
// While this is the closest we have to a programmatic API, please understand that the interfaces in this file are not stable.
// Convert is not included here since it's published independently as @cdktf/hcl2cdk.
/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { Language } from "@cdktf/provider-generator";
export { init, Project, InitArgs } from "./init";
export { get, GetStatus } from "./get";
import { SynthesizedStack } from "./synth-stack";

export { SynthesizedStack, Language };
export { CdktfProject, ProjectUpdate } from "./cdktf-project";
export { watch, State as WatchState } from "./watch";
