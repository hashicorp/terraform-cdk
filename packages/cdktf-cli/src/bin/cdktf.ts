// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
process.env.NODE_ENV = process.env.NODE_ENV || "production";

import * as yargs from "yargs";
import * as path from "path";
import * as os from "os";
import * as fs from "fs-extra";
import * as Sentry from "@sentry/node";
import {
  readCDKTFManifest,
  IsErrorType,
  collectDebugInformation,
  CDKTF_DISABLE_PLUGIN_CACHE_ENV,
} from "@cdktf/commons";

const ensurePluginCache = (): string => {
  const pluginCachePath =
    process.env.TF_PLUGIN_CACHE_DIR ||
    path.join(os.homedir(), ".terraform.d", "plugin-cache");
  if (!fs.existsSync(pluginCachePath)) {
    fs.mkdirpSync(pluginCachePath);
  }
  return pluginCachePath;
};

if (!CDKTF_DISABLE_PLUGIN_CACHE_ENV) {
  process.env.TF_PLUGIN_CACHE_DIR = ensurePluginCache();
}

const customCompletion = function (
  _current: string,
  argv: any,
  completionsFilter: (
    done?: (err: any, defaultCompletions: string[]) => void
  ) => void,
  done: (completions: string[]) => void
) {
  // The completion function gets called when bash completion is invoked and
  // we have very little context there. So when debugging via logs use e.g.
  // fs.appendFileSync("/Users/ansgar/cdktf-completions.log", `hello log\n`);

  // complete positional arg for stack for diff / deploy (and their aliases plan / apply)
  if (
    argv._.includes("diff") ||
    argv._.includes("deploy") ||
    argv._.includes("plan") ||
    argv._.includes("apply")
  ) {
    completionsFilter(async (_err, defaults) => {
      const stacks: string[] = [];
      try {
        const manifest = await readCDKTFManifest();
        stacks.push(
          ...Object.values(manifest.stacks).map(
            (stack) => `${stack.name}:target stack "${stack.name}"`
          )
        );
      } finally {
        done([...defaults, ...stacks]);
      }
    });
  } else {
    completionsFilter();
  }
  // return done([current]);
} as unknown as yargs.AsyncCompletionFunction;
// ^ we need to fake the type here as yargs internally checks whether four
// arguments are present but does not expose the types correctly to typescript
// for the possible overload (which supports falling back to default completions)
// https://github.com/yargs/yargs/blob/d33e9972291406490cd8fdad0b3589be234e0f12/lib/completion.ts#L202

yargs
  .command(require("./cmds/init"))
  .command(require("./cmds/get"))
  .command(require("./cmds/convert"))
  .command(require("./cmds/deploy"))
  .command(require("./cmds/destroy"))
  .command(require("./cmds/diff"))
  .command(require("./cmds/list"))
  .command(require("./cmds/login"))
  .command(require("./cmds/synth"))
  .command(require("./cmds/watch"))
  .command(require("./cmds/output"))
  .command(require("./cmds/debug"))
  .command(require("./cmds/provider"))
  .recommendCommands()
  .exitProcess(false)
  .wrap(yargs.terminalWidth())
  .showHelpOnFail(false)
  .env("CDKTF")
  .epilogue(
    `Options can be specified via environment variables with the "CDKTF_" prefix (e.g. "CDKTF_OUTPUT")`
  )
  .help()
  .alias("h", "help")
  .option("disable-plugin-cache-env", {
    type: "boolean",
    default: false,
    required: false,
    desc: "Dont set TF_PLUGIN_CACHE_DIR automatically. This is useful when the plugin cache is configured differently. Supported using the env CDKTF_DISABLE_PLUGIN_CACHE_ENV.",
  })
  .option("log-level", {
    type: "string",
    required: false,
    desc: "Which log level should be written. Only supported via setting the env CDKTF_LOG_LEVEL",
  })
  .option("context-json", {
    required: false,
    hidden: true,
    desc: "Used internally for env variable",
  })
  .completion("completion", customCompletion) // outputs completion script on "cdktf completion"
  .parserConfiguration({
    "boolean-negation": false,
  })
  .command({
    command: "*", // catches everything not previously matched
    handler: (argv) => {
      const cmd = argv._[0];

      if (cmd === "completion") {
        // Yargs seems to also run the catch all command in case the completion command is run
        return;
      }
      console.error(
        cmd
          ? `Could not find command "${cmd}", here are all available ones:`
          : `Please pass a command to cdktf, here are all available ones:`
      );
      yargs.showHelp();
      process.exit(1);
    },
  })
  .fail(async (message, error) => {
    // will not stop the process, but stops further execution of the handler function
    // this is called first because yargs is not waiting for this async function
    yargs.exit(1, error);

    // set if e.g. the validation of command arguments failed
    if (message) {
      console.log(message);
    }

    // set if e.g. an handler threw an error while being invoked
    if (IsErrorType(error, "Usage") || IsErrorType(error, "External")) {
      console.error(error.message);
    } else if (error) {
      console.error(error.message);
      console.error(error.stack);
      console.error("Collecting Debug Information...");
      const debugOutput = await collectDebugInformation();

      console.error("Debug Information:");
      Object.entries(debugOutput).forEach(([key, value]) => {
        console.log(`${key}: ${value === null ? "null" : value}`);
      });
    }

    await Sentry.close(4000);
    process.exit(1);
  }).argv;
