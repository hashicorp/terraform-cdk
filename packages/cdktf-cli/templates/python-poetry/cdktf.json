{
  "language": "python",
  "app": "poetry run python main.py",
  "terraformProviders": [],
  "terraformModules": [],
  "codeMakerOutput": "imports",
  "context": {
    {{futureFlags}}
  }
}