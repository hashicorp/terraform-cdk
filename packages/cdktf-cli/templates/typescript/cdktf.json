{
  "language": "typescript",
  "app": "node main.js",
  "terraformProviders": ["aws@~> 2.0"]
}