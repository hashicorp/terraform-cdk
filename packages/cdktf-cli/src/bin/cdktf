#!/usr/bin/env node
require("./cdktf.js");
