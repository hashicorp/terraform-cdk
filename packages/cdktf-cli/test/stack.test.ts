import { Node } from "constructs";

import { TerraformResource, TerraformStack, App, Testing } from "cdktf/lib";
import { TerraformModule } from "cdktf/lib/terraform-module";

test('stack synthesis merges all elements into a single output', () => {
  const app = new App();
  const stack = new TerraformStack(app, 'MyStack');

  new MyResource(stack, 'Resource1', {
    terraformResourceType: 'aws_bucket',
    provider: 'my_provider'
  });

  const overrideResource = new MyResource(stack, 'Resource2', {
    terraformResourceType: 'aws_topic',
  })
  overrideResource.addOverride('//', 'this is a comment');
  overrideResource.addOverride('prop2', undefined);
  overrideResource.addOverride('prop3.name', 'test');
  overrideResource.addOverride('provisioner', [{
    'local-exec': {
      command: "echo 'Hello World' >example.txt"
    }
  }]);

  new MyModule(stack, 'EksModule', {
    source: 'terraform-aws-modules/eks/aws',
    version: '7.0.1',
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

class MyModule extends TerraformModule {
  protected synthesizeAttributes() {
    return {
      //eslint-disable-next-line @typescript-eslint/camelcase
      cluster_name: 'my_cluster_name'
    };
  }
}

class MyResource extends TerraformResource {
  protected synthesizeAttributes(): { [name: string]: any } {
    return {
      foo: Node.of(this).uniqueId,
      prop1: 'bar1',
      prop2: 1234,
      prop3: {
        name: 'should be overwritten in resource 2',
        value: 5678
      }
    };
  }
}
