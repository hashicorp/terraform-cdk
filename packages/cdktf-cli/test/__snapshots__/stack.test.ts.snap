// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stack synthesis merges all elements into a single output 1`] = `
Object {
  "module": Object {
    "MyStackEksModule1F7A8254": Object {
      "cluster_name": "my_cluster_name",
      "source": "terraform-aws-modules/eks/aws",
      "version": "7.0.1",
    },
  },
  "resource": Object {
    "aws_bucket": Object {
      "MyStackResource11D44B20C": Object {
        "foo": "MyStackResource11D44B20C",
        "prop1": "bar1",
        "prop2": 1234,
      },
    },
    "aws_topic": Object {
      "MyStackResource2C790E556": Object {
        "foo": "MyStackResource2C790E556",
        "prop1": "bar1",
        "prop2": 1234,
      },
    },
  },
}
`;
